[clang-tidy] Exclude bitwise operators in bugprone-non-zero-enum-to-bool-conversion

PiotrZSL opened the pull request:
Improved bugprone-non-zero-enum-to-bool-conversion check by eliminating false positives resulting from direct usage of bitwise operators.

PiotrZSL committed:
[clang-tidy] Exclude bitwise operators in bugprone-non-zero-enum-to-bool-conversion

Improved bugprone-non-zero-enum-to-bool-conversion check by
eliminating false positives resulting from direct usage of
bitwise operators.

HerrCai0907 commented:
explicit bit usage will not case false positive without this patch. Should we consider case of operator overloading?
```cpp
bool explicitBitUsage1(EStatus value) {
  return (value & SUCCESS);
}
bool explicitBitUsage2(EStatus value) {
  return (value | SUCCESS);
}
```

PiotrZSL commented:
> explicit bit usage will not case false positive without this patch. Should we consider case of operator overloading?

Yes, thats a side effect due to implicit casting enums to integers. I added tests because those were missing, then I implemented version with cxxOverloadCallExpr and changed it into more generic version by using binaryOperation.
If you want, then yes, I can reduce this back to overload operators calls only.

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/NonZeroEnumToBoolConversionCheck.cpp:
```diff
@@ -49,6 +49,9 @@ bool NonZeroEnumToBoolConversionCheck::isLanguageVersionSupported(
 
 void NonZeroEnumToBoolConversionCheck::registerMatchers(MatchFinder *Finder) {
+  auto ExcludedOperators = binaryOperation(hasAnyOperatorName(
+      "|", "&", "^", "<<", ">>", "~", "|=", "&=", "^=", "<<=", ">>="));
```
> If you want, then yes, I can reduce this back to overload operators calls only.

I think current implement it fine, but maybe add some comment here to clarify why you want to exclude those operators. Because actually it's for the operator overloading instead of explicit operator.

PiotrZSL committed:
Add comment
